File: 0149-max-points-on-a-line/0149-max-points-on-a-line.java
import java.util.HashMap;

public class Solution {
    public int maxPoints(int[][] points) {
        if (points.length <= 2) return points.length;

        int max = 1;

        for (int i = 0; i < points.length - 1; i++) {
            HashMap<String, Integer> map = new HashMap<>();
            int localMax = 0;

            for (int j = i + 1; j < points.length; j++) {
                String slope = findSlope(points[i], points[j]);
                map.put(slope, map.getOrDefault(slope, 0) + 1);
                localMax = Math.max(localMax, map.get(slope));
            }

            max = Math.max(max, localMax + 1);

            if (max >= points.length - i) break;
        }

        return max;
    }

      private String findSlope(int[] p1, int[] p2) {
        int x1 = p1[0], y1 = p1[1];
        int x2 = p2[0], y2 = p2[1];
        if (x2 - x1 == 0) return "vert";
        
        if (y2 - y1 == 0) return "horiz";

        return String.valueOf((double)(y2 - y1) / (x2 - x1));
    }
}

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var maxPoints = function(points) {
    if (points.length <= 2) return points.length;

    const gcd = (a, b) => b === 0 ? a : gcd(b, a % b);
    let max = 1;

    for (let i = 0; i < points.length - 1; i++) {
        let map = {};
        let localMax = 0;
let findSlope = function (p1, p2) {
        let [x1,y1] = p1
        let [x2,y2] = p2

        return (y2-y1)/(x2-x1)
    }
        for (let j = i + 1; j < points.length; j++) {
            const slope = findSlope(points[i], points[j])
            map[slope] = (map[slope] || 0) + 1;
            localMax = Math.max(localMax, map[slope]);
        }

        max = Math.max(max, localMax + 1);

        if (max >= points.length - i) break;
    }

    return max;
};
